### PR TITLE
ssecMiddleware migration

### PR DESCRIPTION
**Pull request by srchase**
Migrates and applies ssecMiddlware

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 1]
```diff
@@ -19,18 +19,29 @@
...
 import software.amazon.smithy.typescript.codegen.integration.TypeScriptIntegration;
 import software.amazon.smithy.utils.ListUtils;
+import software.amazon.smithy.utils.SetUtils;
 
 /**
  * Adds all built-in runtime client plugins to clients.
  */
 public class AddBuiltinPlugins implements TypeScriptIntegration {
 
+    private final static Set<String> SSEC_OPERATIONS = SetUtils.of("SSECustomerKey", "CopySourceSSECustomerKey");
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: `private static final`

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AwsDependency.java [hunk 2]
```diff
@@ -34,7 +34,8 @@
     ACCEPTS_HEADER(NORMAL_DEPENDENCY, "@aws-sdk/middleware-sdk-api-gateway", "^0.1.0-preview.5"),
     VALIDATE_BUCKET_NAME(NORMAL_DEPENDENCY, "@aws-sdk/middleware-sdk-s3", "^0.1.0-preview.2"),
     ADD_EXPECT_CONTINUE(NORMAL_DEPENDENCY, "@aws-sdk/middleware-expect-continue", "^0.1.0-preview.5"),
-    ADD_GLACIER_API_VERSION(NORMAL_DEPENDENCY, "@aws-sdk/middleware-sdk-glacier", "^0.1.0-preview.7");
+    ADD_GLACIER_API_VERSION(NORMAL_DEPENDENCY, "@aws-sdk/middleware-sdk-glacier", "^0.1.0-preview.7"),
+    SSEC(NORMAL_DEPENDENCY, "@aws-sdk/middleware-ssec", "^0.1.0-preview.5");
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AwsDependency.java]: This probably should be more description (e.g., `SSEC_MIDDLEWARE`)

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 3]
```diff
@@ -74,10 +85,30 @@
...
+                RuntimeClientPlugin.builder()
+                        .withConventions(AwsDependency.SSEC.dependency, "Ssec", HAS_MIDDLEWARE)
+                        .servicePredicate((m, s) -> testServiceId(s, "S3"))
+                        .operationPredicate((m, s, o) -> testContainsMember(m, o, SSEC_OPERATIONS))
                         .build()
         );
     }
 
+    private static boolean testContainsMember(Model model, OperationShape operationShape, Set<String> expectedMemberNames) {
+        ShapeIndex shapeIndex = model.getShapeIndex();
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: We're deprecating ShapeIndex in favor of just using the Model directly. Any method you were calling on the ShapeIndex should just get called on the Model.

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java [hunk 4]
```diff
@@ -74,10 +85,30 @@
...
+                        .build(),
+                RuntimeClientPlugin.builder()
+                        .withConventions(AwsDependency.SSEC.dependency, "Ssec", HAS_MIDDLEWARE)
+                        .servicePredicate((m, s) -> testServiceId(s, "S3"))
+                        .operationPredicate((m, s, o) -> testContainsMember(m, o, SSEC_OPERATIONS))
                         .build()
         );
     }
 
+    private static boolean testContainsMember(Model model, OperationShape operationShape, Set<String> expectedMemberNames) {
```
Comment (mtdowling) [inline codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: This method can be simplified:

```java
OperationIndex operationIndex = model.getKnowledge(OperationIndex.class);
return operationIndex.getInput(operationShape)
        .filter(input -> input.getMemberNames().stream().anyMatch(expectedMemberNames::contains))
        .isPresent();
```

Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in 6bcea90f30616e16983326a3b87381730520931d

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java (hunk 3 above)
Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in 6bcea90f30616e16983326a3b87381730520931d

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AwsDependency.java (hunk 2 above)
Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AwsDependency.java]: fixed in 6bcea90f30616e16983326a3b87381730520931d

File: codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java (hunk 1 above)
Comment (srchase) [reply codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddBuiltinPlugins.java]: fixed in 6bcea90f30616e16983326a3b87381730520931d

**Pull request merged**